// oci-aispeech/lib/model/customization-model-details.js
"use strict";
/**
 * Speech API
 * The OCI Speech Service harnesses the power of spoken language by allowing developers to easily convert file-based data containing human speech into highly accurate text transcriptions.
 * OpenAPI spec version: 20220101
 *
 *
 * NOTE: This class is auto generated by OracleSDKGenerator.
 * Do not edit the class manually.
 *
 * Copyright (c) 2020, 2026, Oracle and/or its affiliates.  All rights reserved.
 * This software is dual-licensed to you under the Universal Permissive License (UPL) 1.0 as shown at https://oss.oracle.com/licenses/upl or Apache License 2.0 as shown at http://www.apache.org/licenses/LICENSE-2.0. You may choose either license.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomizationModelDetails = void 0;
var CustomizationModelDetails;
(function (CustomizationModelDetails) {
    let Domain;
    (function (Domain) {
        Domain["Generic"] = "GENERIC";
        Domain["Medical"] = "MEDICAL";
        /**
         * This value is used if a service returns a value for this enum that is not recognized by this
         * version of the SDK.
         */
        Domain["UnknownValue"] = "UNKNOWN_VALUE";
    })(Domain = CustomizationModelDetails.Domain || (CustomizationModelDetails.Domain = {}));
    function getJsonObj(obj) {
        const jsonObj = Object.assign(Object.assign({}, obj), {});
        return jsonObj;
    }
    CustomizationModelDetails.getJsonObj = getJsonObj;
    function getDeserializedJsonObj(obj) {
        const jsonObj = Object.assign(Object.assign({}, obj), {});
        return jsonObj;
    }
    CustomizationModelDetails.getDeserializedJsonObj = getDeserializedJsonObj;
})(CustomizationModelDetails = exports.CustomizationModelDetails || (exports.CustomizationModelDetails = {}));
