// oci-aispeech/lib/model/index.js
"use strict";
/**
 * Speech API
 * The OCI Speech Service harnesses the power of spoken language by allowing developers to easily convert file-based data containing human speech into highly accurate text transcriptions.
 * OpenAPI spec version: 20220101
 *
 *
 * NOTE: This class is auto generated by OracleSDKGenerator.
 * Do not edit the class manually.
 *
 * Copyright (c) 2020, 2026, Oracle and/or its affiliates.  All rights reserved.
 * This software is dual-licensed to you under the Universal Permissive License (UPL) 1.0 as shown at https://oss.oracle.com/licenses/upl or Apache License 2.0 as shown at http://www.apache.org/licenses/LICENSE-2.0. You may choose either license.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VoiceCollection = exports.UpdateTranscriptionJobDetails = exports.UpdateCustomizationDetails = exports.TtsOracleSpeechSettings = exports.TtsOracleModelDetails = exports.TtsConfiguration = exports.TtsAudioConfig = exports.TranscriptionTaskSummary = exports.TranscriptionTaskCollection = exports.TranscriptionTask = exports.TranscriptionSettings = exports.TranscriptionNormalization = exports.TranscriptionModelDetails = exports.TranscriptionJobSummary = exports.TranscriptionJobCollection = exports.TranscriptionJob = exports.TranscriptionFilter = exports.SynthesizeSpeechDetails = exports.SortOrder = exports.RealtimeSessionToken = exports.RealtimeParameters = exports.RealtimeModelDetails = exports.RealtimeMessageResultTranscriptionToken = exports.RealtimeMessageResultTranscription = exports.RealtimeMessageOut = exports.RealtimeMessageAuthentication = exports.RealtimeMessageAckAudioDetails = exports.RealtimeMessage = exports.Pronunciation = exports.OutputLocation = exports.ObjectLocation = exports.LocationDetails = exports.InputLocation = exports.EntityList = exports.Entity = exports.Diarization = exports.CustomizationSummary = exports.CustomizationReferencedEntities = exports.CustomizationModelDetails = exports.CustomizationInferenceEntity = exports.CustomizationInference = exports.CustomizationDatasetDetails = exports.CustomizationCollection = exports.Customization = exports.CreateTranscriptionJobDetails = exports.CreateRealtimeSessionTokenDetails = exports.CreateCustomizationDetails = exports.ChangeTranscriptionJobCompartmentDetails = exports.ChangeCustomizationCompartmentDetails = exports.AudioFormatDetails = void 0;
exports.TtsOracleTts2NaturalModelDetails = exports.TtsOracleTts1StandardModelDetails = exports.TtsOracleConfiguration = exports.TtsBaseAudioConfig = exports.RealtimeMessageSendFinalResult = exports.RealtimeMessageResult = exports.RealtimeMessageError = exports.RealtimeMessageConnect = exports.RealtimeMessageAuthenticationToken = exports.RealtimeMessageAuthenticationCredentials = exports.RealtimeMessageAckAudio = exports.ProfanityTranscriptionFilter = exports.ObjectStorageDataset = exports.ObjectListInlineInputLocation = exports.ObjectListFileInputLocation = exports.ObjectListDataset = exports.EntityListDataset = exports.VoiceSummary = void 0;
const AudioFormatDetails = __importStar(require("./audio-format-details"));
exports.AudioFormatDetails = AudioFormatDetails.AudioFormatDetails;
const ChangeCustomizationCompartmentDetails = __importStar(require("./change-customization-compartment-details"));
exports.ChangeCustomizationCompartmentDetails = ChangeCustomizationCompartmentDetails.ChangeCustomizationCompartmentDetails;
const ChangeTranscriptionJobCompartmentDetails = __importStar(require("./change-transcription-job-compartment-details"));
exports.ChangeTranscriptionJobCompartmentDetails = ChangeTranscriptionJobCompartmentDetails.ChangeTranscriptionJobCompartmentDetails;
const CreateCustomizationDetails = __importStar(require("./create-customization-details"));
exports.CreateCustomizationDetails = CreateCustomizationDetails.CreateCustomizationDetails;
const CreateRealtimeSessionTokenDetails = __importStar(require("./create-realtime-session-token-details"));
exports.CreateRealtimeSessionTokenDetails = CreateRealtimeSessionTokenDetails.CreateRealtimeSessionTokenDetails;
const CreateTranscriptionJobDetails = __importStar(require("./create-transcription-job-details"));
exports.CreateTranscriptionJobDetails = CreateTranscriptionJobDetails.CreateTranscriptionJobDetails;
const Customization = __importStar(require("./customization"));
exports.Customization = Customization.Customization;
const CustomizationCollection = __importStar(require("./customization-collection"));
exports.CustomizationCollection = CustomizationCollection.CustomizationCollection;
const CustomizationDatasetDetails = __importStar(require("./customization-dataset-details"));
exports.CustomizationDatasetDetails = CustomizationDatasetDetails.CustomizationDatasetDetails;
const CustomizationInference = __importStar(require("./customization-inference"));
exports.CustomizationInference = CustomizationInference.CustomizationInference;
const CustomizationInferenceEntity = __importStar(require("./customization-inference-entity"));
exports.CustomizationInferenceEntity = CustomizationInferenceEntity.CustomizationInferenceEntity;
const CustomizationModelDetails = __importStar(require("./customization-model-details"));
exports.CustomizationModelDetails = CustomizationModelDetails.CustomizationModelDetails;
const CustomizationReferencedEntities = __importStar(require("./customization-referenced-entities"));
exports.CustomizationReferencedEntities = CustomizationReferencedEntities.CustomizationReferencedEntities;
const CustomizationSummary = __importStar(require("./customization-summary"));
exports.CustomizationSummary = CustomizationSummary.CustomizationSummary;
const Diarization = __importStar(require("./diarization"));
exports.Diarization = Diarization.Diarization;
const Entity = __importStar(require("./entity"));
exports.Entity = Entity.Entity;
const EntityList = __importStar(require("./entity-list"));
exports.EntityList = EntityList.EntityList;
const InputLocation = __importStar(require("./input-location"));
exports.InputLocation = InputLocation.InputLocation;
const LocationDetails = __importStar(require("./location-details"));
exports.LocationDetails = LocationDetails.LocationDetails;
const ObjectLocation = __importStar(require("./object-location"));
exports.ObjectLocation = ObjectLocation.ObjectLocation;
const OutputLocation = __importStar(require("./output-location"));
exports.OutputLocation = OutputLocation.OutputLocation;
const Pronunciation = __importStar(require("./pronunciation"));
exports.Pronunciation = Pronunciation.Pronunciation;
const RealtimeMessage = __importStar(require("./realtime-message"));
exports.RealtimeMessage = RealtimeMessage.RealtimeMessage;
const RealtimeMessageAckAudioDetails = __importStar(require("./realtime-message-ack-audio-details"));
exports.RealtimeMessageAckAudioDetails = RealtimeMessageAckAudioDetails.RealtimeMessageAckAudioDetails;
const RealtimeMessageAuthentication = __importStar(require("./realtime-message-authentication"));
exports.RealtimeMessageAuthentication = RealtimeMessageAuthentication.RealtimeMessageAuthentication;
const RealtimeMessageOut = __importStar(require("./realtime-message-out"));
exports.RealtimeMessageOut = RealtimeMessageOut.RealtimeMessageOut;
const RealtimeMessageResultTranscription = __importStar(require("./realtime-message-result-transcription"));
exports.RealtimeMessageResultTranscription = RealtimeMessageResultTranscription.RealtimeMessageResultTranscription;
const RealtimeMessageResultTranscriptionToken = __importStar(require("./realtime-message-result-transcription-token"));
exports.RealtimeMessageResultTranscriptionToken = RealtimeMessageResultTranscriptionToken.RealtimeMessageResultTranscriptionToken;
const RealtimeModelDetails = __importStar(require("./realtime-model-details"));
exports.RealtimeModelDetails = RealtimeModelDetails.RealtimeModelDetails;
const RealtimeParameters = __importStar(require("./realtime-parameters"));
exports.RealtimeParameters = RealtimeParameters.RealtimeParameters;
const RealtimeSessionToken = __importStar(require("./realtime-session-token"));
exports.RealtimeSessionToken = RealtimeSessionToken.RealtimeSessionToken;
const SortOrder = __importStar(require("./sort-order"));
exports.SortOrder = SortOrder.SortOrder;
const SynthesizeSpeechDetails = __importStar(require("./synthesize-speech-details"));
exports.SynthesizeSpeechDetails = SynthesizeSpeechDetails.SynthesizeSpeechDetails;
const TranscriptionFilter = __importStar(require("./transcription-filter"));
exports.TranscriptionFilter = TranscriptionFilter.TranscriptionFilter;
const TranscriptionJob = __importStar(require("./transcription-job"));
exports.TranscriptionJob = TranscriptionJob.TranscriptionJob;
const TranscriptionJobCollection = __importStar(require("./transcription-job-collection"));
exports.TranscriptionJobCollection = TranscriptionJobCollection.TranscriptionJobCollection;
const TranscriptionJobSummary = __importStar(require("./transcription-job-summary"));
exports.TranscriptionJobSummary = TranscriptionJobSummary.TranscriptionJobSummary;
const TranscriptionModelDetails = __importStar(require("./transcription-model-details"));
exports.TranscriptionModelDetails = TranscriptionModelDetails.TranscriptionModelDetails;
const TranscriptionNormalization = __importStar(require("./transcription-normalization"));
exports.TranscriptionNormalization = TranscriptionNormalization.TranscriptionNormalization;
const TranscriptionSettings = __importStar(require("./transcription-settings"));
exports.TranscriptionSettings = TranscriptionSettings.TranscriptionSettings;
const TranscriptionTask = __importStar(require("./transcription-task"));
exports.TranscriptionTask = TranscriptionTask.TranscriptionTask;
const TranscriptionTaskCollection = __importStar(require("./transcription-task-collection"));
exports.TranscriptionTaskCollection = TranscriptionTaskCollection.TranscriptionTaskCollection;
const TranscriptionTaskSummary = __importStar(require("./transcription-task-summary"));
exports.TranscriptionTaskSummary = TranscriptionTaskSummary.TranscriptionTaskSummary;
const TtsAudioConfig = __importStar(require("./tts-audio-config"));
exports.TtsAudioConfig = TtsAudioConfig.TtsAudioConfig;
const TtsConfiguration = __importStar(require("./tts-configuration"));
exports.TtsConfiguration = TtsConfiguration.TtsConfiguration;
const TtsOracleModelDetails = __importStar(require("./tts-oracle-model-details"));
exports.TtsOracleModelDetails = TtsOracleModelDetails.TtsOracleModelDetails;
const TtsOracleSpeechSettings = __importStar(require("./tts-oracle-speech-settings"));
exports.TtsOracleSpeechSettings = TtsOracleSpeechSettings.TtsOracleSpeechSettings;
const UpdateCustomizationDetails = __importStar(require("./update-customization-details"));
exports.UpdateCustomizationDetails = UpdateCustomizationDetails.UpdateCustomizationDetails;
const UpdateTranscriptionJobDetails = __importStar(require("./update-transcription-job-details"));
exports.UpdateTranscriptionJobDetails = UpdateTranscriptionJobDetails.UpdateTranscriptionJobDetails;
const VoiceCollection = __importStar(require("./voice-collection"));
exports.VoiceCollection = VoiceCollection.VoiceCollection;
const VoiceSummary = __importStar(require("./voice-summary"));
exports.VoiceSummary = VoiceSummary.VoiceSummary;
const EntityListDataset = __importStar(require("./entity-list-dataset"));
exports.EntityListDataset = EntityListDataset.EntityListDataset;
const ObjectListDataset = __importStar(require("./object-list-dataset"));
exports.ObjectListDataset = ObjectListDataset.ObjectListDataset;
const ObjectListFileInputLocation = __importStar(require("./object-list-file-input-location"));
exports.ObjectListFileInputLocation = ObjectListFileInputLocation.ObjectListFileInputLocation;
const ObjectListInlineInputLocation = __importStar(require("./object-list-inline-input-location"));
exports.ObjectListInlineInputLocation = ObjectListInlineInputLocation.ObjectListInlineInputLocation;
const ObjectStorageDataset = __importStar(require("./object-storage-dataset"));
exports.ObjectStorageDataset = ObjectStorageDataset.ObjectStorageDataset;
const ProfanityTranscriptionFilter = __importStar(require("./profanity-transcription-filter"));
exports.ProfanityTranscriptionFilter = ProfanityTranscriptionFilter.ProfanityTranscriptionFilter;
const RealtimeMessageAckAudio = __importStar(require("./realtime-message-ack-audio"));
exports.RealtimeMessageAckAudio = RealtimeMessageAckAudio.RealtimeMessageAckAudio;
const RealtimeMessageAuthenticationCredentials = __importStar(require("./realtime-message-authentication-credentials"));
exports.RealtimeMessageAuthenticationCredentials = RealtimeMessageAuthenticationCredentials.RealtimeMessageAuthenticationCredentials;
const RealtimeMessageAuthenticationToken = __importStar(require("./realtime-message-authentication-token"));
exports.RealtimeMessageAuthenticationToken = RealtimeMessageAuthenticationToken.RealtimeMessageAuthenticationToken;
const RealtimeMessageConnect = __importStar(require("./realtime-message-connect"));
exports.RealtimeMessageConnect = RealtimeMessageConnect.RealtimeMessageConnect;
const RealtimeMessageError = __importStar(require("./realtime-message-error"));
exports.RealtimeMessageError = RealtimeMessageError.RealtimeMessageError;
const RealtimeMessageResult = __importStar(require("./realtime-message-result"));
exports.RealtimeMessageResult = RealtimeMessageResult.RealtimeMessageResult;
const RealtimeMessageSendFinalResult = __importStar(require("./realtime-message-send-final-result"));
exports.RealtimeMessageSendFinalResult = RealtimeMessageSendFinalResult.RealtimeMessageSendFinalResult;
const TtsBaseAudioConfig = __importStar(require("./tts-base-audio-config"));
exports.TtsBaseAudioConfig = TtsBaseAudioConfig.TtsBaseAudioConfig;
const TtsOracleConfiguration = __importStar(require("./tts-oracle-configuration"));
exports.TtsOracleConfiguration = TtsOracleConfiguration.TtsOracleConfiguration;
const TtsOracleTts1StandardModelDetails = __importStar(require("./tts-oracle-tts1-standard-model-details"));
exports.TtsOracleTts1StandardModelDetails = TtsOracleTts1StandardModelDetails.TtsOracleTts1StandardModelDetails;
const TtsOracleTts2NaturalModelDetails = __importStar(require("./tts-oracle-tts2-natural-model-details"));
exports.TtsOracleTts2NaturalModelDetails = TtsOracleTts2NaturalModelDetails.TtsOracleTts2NaturalModelDetails;
