// oci-aispeech/lib/model/profanity-transcription-filter.js
"use strict";
/**
 * Speech API
 * The OCI Speech Service harnesses the power of spoken language by allowing developers to easily convert file-based data containing human speech into highly accurate text transcriptions.
 * OpenAPI spec version: 20220101
 *
 *
 * NOTE: This class is auto generated by OracleSDKGenerator.
 * Do not edit the class manually.
 *
 * Copyright (c) 2020, 2026, Oracle and/or its affiliates.  All rights reserved.
 * This software is dual-licensed to you under the Universal Permissive License (UPL) 1.0 as shown at https://oss.oracle.com/licenses/upl or Apache License 2.0 as shown at http://www.apache.org/licenses/LICENSE-2.0. You may choose either license.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProfanityTranscriptionFilter = void 0;
const model = __importStar(require("../model"));
var ProfanityTranscriptionFilter;
(function (ProfanityTranscriptionFilter) {
    let Mode;
    (function (Mode) {
        Mode["Mask"] = "MASK";
        Mode["Remove"] = "REMOVE";
        Mode["Tag"] = "TAG";
        /**
         * This value is used if a service returns a value for this enum that is not recognized by this
         * version of the SDK.
         */
        Mode["UnknownValue"] = "UNKNOWN_VALUE";
    })(Mode = ProfanityTranscriptionFilter.Mode || (ProfanityTranscriptionFilter.Mode = {}));
    function getJsonObj(obj, isParentJsonObj) {
        const jsonObj = Object.assign(Object.assign({}, (isParentJsonObj
            ? obj
            : model.TranscriptionFilter.getJsonObj(obj))), {});
        return jsonObj;
    }
    ProfanityTranscriptionFilter.getJsonObj = getJsonObj;
    ProfanityTranscriptionFilter.type = "PROFANITY";
    function getDeserializedJsonObj(obj, isParentJsonObj) {
        const jsonObj = Object.assign(Object.assign({}, (isParentJsonObj
            ? obj
            : model.TranscriptionFilter.getDeserializedJsonObj(obj))), {});
        return jsonObj;
    }
    ProfanityTranscriptionFilter.getDeserializedJsonObj = getDeserializedJsonObj;
})(ProfanityTranscriptionFilter = exports.ProfanityTranscriptionFilter || (exports.ProfanityTranscriptionFilter = {}));
