// oci-aispeech/lib/model/realtime-message-result-transcription-token.js
"use strict";
/**
 * Speech API
 * The OCI Speech Service harnesses the power of spoken language by allowing developers to easily convert file-based data containing human speech into highly accurate text transcriptions.
 * OpenAPI spec version: 20220101
 *
 *
 * NOTE: This class is auto generated by OracleSDKGenerator.
 * Do not edit the class manually.
 *
 * Copyright (c) 2020, 2026, Oracle and/or its affiliates.  All rights reserved.
 * This software is dual-licensed to you under the Universal Permissive License (UPL) 1.0 as shown at https://oss.oracle.com/licenses/upl or Apache License 2.0 as shown at http://www.apache.org/licenses/LICENSE-2.0. You may choose either license.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RealtimeMessageResultTranscriptionToken = void 0;
var RealtimeMessageResultTranscriptionToken;
(function (RealtimeMessageResultTranscriptionToken) {
    let Type;
    (function (Type) {
        Type["Word"] = "WORD";
        Type["Punctuation"] = "PUNCTUATION";
    })(Type = RealtimeMessageResultTranscriptionToken.Type || (RealtimeMessageResultTranscriptionToken.Type = {}));
    function getJsonObj(obj) {
        const jsonObj = Object.assign(Object.assign({}, obj), {});
        return jsonObj;
    }
    RealtimeMessageResultTranscriptionToken.getJsonObj = getJsonObj;
    function getDeserializedJsonObj(obj) {
        const jsonObj = Object.assign(Object.assign({}, obj), {});
        return jsonObj;
    }
    RealtimeMessageResultTranscriptionToken.getDeserializedJsonObj = getDeserializedJsonObj;
})(RealtimeMessageResultTranscriptionToken = exports.RealtimeMessageResultTranscriptionToken || (exports.RealtimeMessageResultTranscriptionToken = {}));
