// oci-aispeech/lib/model/transcription-task.js
"use strict";
/**
 * Speech API
 * The OCI Speech Service harnesses the power of spoken language by allowing developers to easily convert file-based data containing human speech into highly accurate text transcriptions.
 * OpenAPI spec version: 20220101
 *
 *
 * NOTE: This class is auto generated by OracleSDKGenerator.
 * Do not edit the class manually.
 *
 * Copyright (c) 2020, 2026, Oracle and/or its affiliates.  All rights reserved.
 * This software is dual-licensed to you under the Universal Permissive License (UPL) 1.0 as shown at https://oss.oracle.com/licenses/upl or Apache License 2.0 as shown at http://www.apache.org/licenses/LICENSE-2.0. You may choose either license.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TranscriptionTask = void 0;
const model = __importStar(require("../model"));
var TranscriptionTask;
(function (TranscriptionTask) {
    let LifecycleState;
    (function (LifecycleState) {
        LifecycleState["Accepted"] = "ACCEPTED";
        LifecycleState["InProgress"] = "IN_PROGRESS";
        LifecycleState["Succeeded"] = "SUCCEEDED";
        LifecycleState["Failed"] = "FAILED";
        LifecycleState["Canceled"] = "CANCELED";
        /**
         * This value is used if a service returns a value for this enum that is not recognized by this
         * version of the SDK.
         */
        LifecycleState["UnknownValue"] = "UNKNOWN_VALUE";
    })(LifecycleState = TranscriptionTask.LifecycleState || (TranscriptionTask.LifecycleState = {}));
    function getJsonObj(obj) {
        const jsonObj = Object.assign(Object.assign({}, obj), {
            "modelDetails": obj.modelDetails
                ? model.TranscriptionModelDetails.getJsonObj(obj.modelDetails)
                : undefined,
            "audioFormatDetails": obj.audioFormatDetails
                ? model.AudioFormatDetails.getJsonObj(obj.audioFormatDetails)
                : undefined,
            "inputLocation": obj.inputLocation
                ? model.ObjectLocation.getJsonObj(obj.inputLocation)
                : undefined,
            "outputLocation": obj.outputLocation
                ? model.ObjectLocation.getJsonObj(obj.outputLocation)
                : undefined
        });
        return jsonObj;
    }
    TranscriptionTask.getJsonObj = getJsonObj;
    function getDeserializedJsonObj(obj) {
        const jsonObj = Object.assign(Object.assign({}, obj), {
            "modelDetails": obj.modelDetails
                ? model.TranscriptionModelDetails.getDeserializedJsonObj(obj.modelDetails)
                : undefined,
            "audioFormatDetails": obj.audioFormatDetails
                ? model.AudioFormatDetails.getDeserializedJsonObj(obj.audioFormatDetails)
                : undefined,
            "inputLocation": obj.inputLocation
                ? model.ObjectLocation.getDeserializedJsonObj(obj.inputLocation)
                : undefined,
            "outputLocation": obj.outputLocation
                ? model.ObjectLocation.getDeserializedJsonObj(obj.outputLocation)
                : undefined
        });
        return jsonObj;
    }
    TranscriptionTask.getDeserializedJsonObj = getDeserializedJsonObj;
})(TranscriptionTask = exports.TranscriptionTask || (exports.TranscriptionTask = {}));
