// oci-aispeech/lib/model/tts-oracle-speech-settings.js
"use strict";
/**
 * Speech API
 * The OCI Speech Service harnesses the power of spoken language by allowing developers to easily convert file-based data containing human speech into highly accurate text transcriptions.
 * OpenAPI spec version: 20220101
 *
 *
 * NOTE: This class is auto generated by OracleSDKGenerator.
 * Do not edit the class manually.
 *
 * Copyright (c) 2020, 2026, Oracle and/or its affiliates.  All rights reserved.
 * This software is dual-licensed to you under the Universal Permissive License (UPL) 1.0 as shown at https://oss.oracle.com/licenses/upl or Apache License 2.0 as shown at http://www.apache.org/licenses/LICENSE-2.0. You may choose either license.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TtsOracleSpeechSettings = void 0;
var TtsOracleSpeechSettings;
(function (TtsOracleSpeechSettings) {
    let TextType;
    (function (TextType) {
        TextType["Text"] = "TEXT";
        TextType["Ssml"] = "SSML";
    })(TextType = TtsOracleSpeechSettings.TextType || (TtsOracleSpeechSettings.TextType = {}));
    let OutputFormat;
    (function (OutputFormat) {
        OutputFormat["Mp3"] = "MP3";
        OutputFormat["Ogg"] = "OGG";
        OutputFormat["Pcm"] = "PCM";
        OutputFormat["Json"] = "JSON";
    })(OutputFormat = TtsOracleSpeechSettings.OutputFormat || (TtsOracleSpeechSettings.OutputFormat = {}));
    let SpeechMarkTypes;
    (function (SpeechMarkTypes) {
        SpeechMarkTypes["Sentence"] = "SENTENCE";
        SpeechMarkTypes["Word"] = "WORD";
    })(SpeechMarkTypes = TtsOracleSpeechSettings.SpeechMarkTypes || (TtsOracleSpeechSettings.SpeechMarkTypes = {}));
    function getJsonObj(obj) {
        const jsonObj = Object.assign(Object.assign({}, obj), {});
        return jsonObj;
    }
    TtsOracleSpeechSettings.getJsonObj = getJsonObj;
    function getDeserializedJsonObj(obj) {
        const jsonObj = Object.assign(Object.assign({}, obj), {});
        return jsonObj;
    }
    TtsOracleSpeechSettings.getDeserializedJsonObj = getDeserializedJsonObj;
})(TtsOracleSpeechSettings = exports.TtsOracleSpeechSettings || (exports.TtsOracleSpeechSettings = {}));
